Source/WebCore/dom/InternalObserverInspect.cpp

```diff
@@ -0,0 +1,252 @@
+            : SubscriberCallback(&context)
+            , m_sourceObservable(source)
+            , m_inspector(inspector)
+        { }
+
+        Ref<Observable> m_sourceObservable;
+        ObservableInspector m_inspector;
+    };
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        if (m_inspector.next) {
+            Ref vm = protectedGlobalObjectVM();
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.next->handleEventRethrowingException(value);
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                m_subscriber->error(exception->value());
+                return;
+            }
+        }
+
+        m_subscriber->next(value);
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        removeAbortHandler();
+
+        if (m_inspector.error) {
+            Ref vm = protectedGlobalObjectVM();
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.error->handleEventRethrowingException(value);
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                m_subscriber->error(exception->value());
+                return;
+            }
+        }
+
+        m_subscriber->error(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+
+        removeAbortHandler();
+
+        if (m_inspector.complete) {
+            Ref vm = protectedGlobalObjectVM();
+            JSC::JSLockHolder lock(vm);
+            auto scope = DECLARE_CATCH_SCOPE(vm);
+
+            m_inspector.complete->handleEventRethrowingException();
+
+            JSC::Exception* exception = scope.exception();
+            if (UNLIKELY(exception)) {
+                scope.clearException();
+                m_subscriber->error(exception->value());
+                return;
+            }
+        }
+
+        m_subscriber->complete();
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor& visitor) const final
+    {
+        m_subscriber->visitAdditionalChildren(visitor);
+        if (m_inspector.next)
+            m_inspector.next->visitJSFunction(visitor);
+        if (m_inspector.error)
+            m_inspector.error->visitJSFunction(visitor);
+        if (m_inspector.complete)
+            m_inspector.complete->visitJSFunction(visitor);
+        if (m_inspector.subscribe)
+            m_inspector.subscribe->visitJSFunction(visitor);
+        if (m_inspector.abort)
+            m_inspector.abort->visitJSFunction(visitor);
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor& visitor) const final
+    {
+        m_subscriber->visitAdditionalChildren(visitor);
```

Comment:
We should avoid code duplication between the 2 visitAdditionalChildren() functions and have them call a templated function (with the visitor type as template parameter).

Reply:
As mentioned over Slack, i've gone with a templated method today, and in a second iteration we will remove the upstream 2nd method, given the AbstractSlotVisitor